Trace directly into first-class module form.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19694 Remove duplicate inlineCallToCode
* #19603 Serialize first-class version of functions
* **#19371 Trace directly into first-class module form.**
* #19334 @torch.jit.script(fn) now is a torch.jit.Function
* #19333 pybind CompilationUnit and Function directly

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

Differential Revision: [D15016374](https://our.internmc.facebook.com/intern/diff/D15016374)

zdevito committed:
[WIP] Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

zdevito committed:
Update on "[WIP] Trace directly into first-class module form."

zdevito committed:
Update on "[WIP] Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "[WIP] Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

suo reviewed (approved):
Lgtm. My only comment is to add some documentation for some of the newly introduced functions in core places, like remapTypes in ir.h

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "@torch.jit.script(fn) now is a torch.jit.Function"

@torch.jit.script(fn) now is a torch.jit.Function

gh-metadata: pytorch pytorch 19334 gh/zdevito/22/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update base for Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head

zdevito committed:
Update on "Trace directly into first-class module form."

Trace directly into first-class module form.

The trace logic now inserts a `self` argument rather than
additional inputs.

With this change, we no longer have to _define_ methods in
lowered form (with additional Slot inputs).

We still continue to lower methods before execution.

Now that we only execute methods in lowered form, it is possible
to remove the lowered_methods() table. Instead, each individual
method creates its lowered Function on creation.

Methods are still lazily created but there is no additional
need to keep lowered_methods() and the class methods in sync.

gh-metadata: pytorch pytorch 19371 gh/zdevito/23/head
